### PR TITLE
MTDSA-4318 Request Parser and Validation

### PR DESCRIPTION
**Pull request by Nboaram**
don't merge this until after 4316 just in case it causes a problem.

### CODE REVIEW

File: app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import v1.models.des.IncomeSourceType
+import v1.models.domain.TypeOfBusiness
+import v1.models.requestData.{DesTaxYear, RetrieveUKPropertyBISSRawData, RetrieveUKPropertyBISSRequest}
+
+class RetrieveUKPropertyBISSRequestDataParser @Inject()(val validator: RetrieveUKPropertyBISSValidator)
+  extends RequestParser[RetrieveUKPropertyBISSRawData, RetrieveUKPropertyBISSRequest] {
+
+  override protected def requestFor(data: RetrieveUKPropertyBISSRawData): RetrieveUKPropertyBISSRequest = {
+    RetrieveUKPropertyBISSRequest(Nino(data.nino),
+      data.taxYear.map(DesTaxYear.fromMtd),
```
Comment (MarkAKelly) [inline app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala]: Naren has added a set of functions which can be found in `DateUtils` which populate this parameter differently.  We want to default this to the current tax year here if no date is supplied.   Check out `RetrieveSelfEmploymentBISSRequestParser`

Comment (MarkAKelly) [reply app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala]: (and the associated tests)

Comment (Nboaram) [reply app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala]: This should now be defaulting?

File: app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import v1.controllers.requestParsers.validators.RetrieveUKPropertyBISSValidator
+import v1.models.des.IncomeSourceType
+import v1.models.requestData.{RetrieveUKPropertyBISSRawData, RetrieveUKPropertyBISSRequest}
+
+class RetrieveUKPropertyBISSRequestDataParser @Inject()(val validator: RetrieveUKPropertyBISSValidator)
+  extends RequestParser[RetrieveUKPropertyBISSRawData, RetrieveUKPropertyBISSRequest] {
+
+  override protected def requestFor(data: RetrieveUKPropertyBISSRawData): RetrieveUKPropertyBISSRequest = {
+    RetrieveUKPropertyBISSRequest(Nino(data.nino),
+      data.taxYear.fold(DateUtils.getDesTaxYear(LocalDate.now()))(DateUtils.getDesTaxYear),
```
Comment (MarkAKelly) [inline app/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestDataParser.scala]: This is good, just needs a corresponding test.

**Comment by Nboaram** [comment]
This has been rebased again, should now be ready for merging.

File: test/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestParserSpec.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    lazy val parser = new RetrieveUKPropertyBISSRequestDataParser(mockValidator)
+  }
+
+
+  "parse" should {
+    "return a request object" when {
+      "valid data is provided" in new Test {
+        MockValidator.validate(inputData).returns(Nil)
+
+        parser.parseRequest(inputData) shouldBe Right(RetrieveUKPropertyBISSRequest(Nino(nino), DesTaxYear.fromMtd(taxYear), typeOfBusiness match {
```
Comment (james-work-account) [inline test/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestParserSpec.scala]: Why do this pattern matching when you know you're setting typeOfBusiness to `uk-property-non-fhl`?

File: test/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestParserSpec.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          case "uk-property-fhl" => IncomeSourceType.`fhl-property-uk`
+          case "uk-property-non-fhl" => IncomeSourceType.`uk-property`
+        }))
+      }
+    }
+
+    "valid data is provided without tax year" in new Test {
+      MockValidator.validate(inputData.copy(taxYear = None)).returns(Nil)
+
+      parser.parseRequest(inputData.copy(taxYear = None)) shouldBe Right(RetrieveUKPropertyBISSRequest(Nino(nino), DateUtils.getDesTaxYear(LocalDate.now()), typeOfBusiness match {
```
Comment (james-work-account) [inline test/v1/controllers/requestParsers/RetrieveUKPropertyBISSRequestParserSpec.scala]: Why do this pattern matching when you know you're setting typeOfBusiness to `uk-property-non-fhl`?

File: test/v1/controllers/requestParsers/validators/validations/UkPropertyTypeOfBusinessValidationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import v1.models.errors.TypeOfBusinessFormatError
+
+class UkPropertyTypeOfBusinessValidationSpec extends UnitSpec{
+
+  private val typeOfBusiness = "uk-property-fhl"
+  private val badTypeOfBusiness = "rabbit hair comb shop"
+  "validate" should {
+    "return no errors" when {
+      "a valid type of business is provided" in {
+        UKPropertyTypeOfBusinessValidation.validate(typeOfBusiness) shouldBe Nil
```
Comment (james-work-account) [inline test/v1/controllers/requestParsers/validators/validations/UkPropertyTypeOfBusinessValidationSpec.scala]: Need test for `uk-property-non-fhl` too

**Pull request merged**